File: src/TaskOTime/TaskOTime.ViewModel/ViewModels/VmMain.vb
Imports System
Imports System.Collections.ObjectModel
Imports System.ComponentModel
Imports System.Windows.Input
Imports TaskOTime.ViewModel.Base

Namespace ViewModels
    Public Class VmMain
        Inherits ViewModelBase

        Private _selectedDate As DateTime
        Private _isWideLayout As Boolean = True
        Private _manualCompletionStartText As String = DateTime.Now.ToString("HH:mm")
        Private _manualCompletionDurationText As String = "00:30"

        Public Sub New(Optional suppliedTimeCollection As TimeCollectionViewModel = Nothing,
                       Optional suppliedTaskManagement As TaskManagementViewModel = Nothing)
            _selectedDate = DateTime.Today
            Options = New AppOptionsViewModel()
            Me.TimeCollection = If(suppliedTimeCollection, New TimeCollectionViewModel(_selectedDate))
            _selectedDate = Me.TimeCollection.BookingDate
            AddHandler TimeCollection.PropertyChanged, AddressOf OnTimeCollectionPropertyChanged
            AddHandler TimeCollection.TimeEntryCreated, AddressOf OnTimeEntryCreated
            AddHandler TimeCollection.TimeEntryEditRequested, AddressOf OnTimeEntryEditRequested

            Me.TaskManagement = If(suppliedTaskManagement, New TaskManagementViewModel())
            AddHandler TaskManagement.PropertyChanged, AddressOf OnTaskManagementPropertyChanged
            AddHandler TaskManagement.TaskListEditRequested, AddressOf OnTaskListEditRequested
            AddHandler TaskManagement.TaskCompletionRequested, AddressOf OnTaskCompletionRequested

            TodayCommand = New DelegateCommand(AddressOf SelectToday)
            YesterdayCommand = New DelegateCommand(AddressOf SelectYesterday)
            OptionsCommand = New DelegateCommand(AddressOf ShowOptionsDialog)
            ExportSelectedDayCommand = New DelegateCommand(AddressOf ShowExportSelectedDayDialog)
            ExportPeriodCommand = New DelegateCommand(AddressOf ShowExportPeriodDialog)
            ManageProjectsCommand = New DelegateCommand(AddressOf ShowProjectsDialog)
            ManageTaskListsCommand = New DelegateCommand(AddressOf ShowTaskListsDialog)
            ManageTasksCommand = New DelegateCommand(AddressOf ShowTasksDialog)
            ManageUsersAdminCommand = New DelegateCommand(AddressOf ShowUsersAdminDialog)
            ShowDailyStatementCommand = New DelegateCommand(AddressOf ShowDailyStatementDialog)
            ShowWeeklyStatementCommand = New DelegateCommand(AddressOf ShowWeeklyStatementDialog)
            ShowMonthlyStatementCommand = New DelegateCommand(AddressOf ShowMonthlyStatementDialog)
            ShowTenantAdminStatisticsCommand = New DelegateCommand(AddressOf ShowTenantAdminStatisticsDialog)
        End Sub

        Public Event DialogRequested As EventHandler(Of DialogRequestedEventArgs)
        Public Event OptionsRequested As EventHandler(Of AppOptionsViewModel)
        Public Event TaskListEditRequested As EventHandler(Of TaskListEditRequestEventArgs)
        Public Event TimeEntryEditRequested As EventHandler(Of TimeEntryEditRequestEventArgs)
        Public Event MasterDataRequested As EventHandler(Of Integer)

        Public Property SelectedDate As DateTime
            Get
                Return _selectedDate
            End Get
            Set(value As DateTime)
                If SetProperty(_selectedDate, value.Date, NameOf(SelectedDate)) Then
                    TimeCollection.BookingDate = value.Date
                    OnPropertyChanged(NameOf(SelectedDateSummary))
                End If
            End Set
        End Property

        Public Property IsWideLayout As Boolean
            Get
                Return _isWideLayout
            End Get
            Set(value As Boolean)
                SetProperty(_isWideLayout, value, NameOf(IsWideLayout))
            End Set
        End Property

        Public ReadOnly Property TimeCollection As TimeCollectionViewModel

        Public ReadOnly Property BookedDates As ObservableCollection(Of BookedDateItemViewModel)
            Get
                Return TimeCollection.BookedDateItems
            End Get
        End Property

        Public ReadOnly Property TaskManagement As TaskManagementViewModel

        Public ReadOnly Property Options As AppOptionsViewModel

        Public ReadOnly Property TodayCommand As ICommand

        Public ReadOnly Property YesterdayCommand As ICommand

        Public ReadOnly Property OptionsCommand As ICommand

        Public ReadOnly Property ExportSelectedDayCommand As ICommand

        Public ReadOnly Property ExportPeriodCommand As ICommand

        Public ReadOnly Property ManageProjectsCommand As ICommand

        Public ReadOnly Property ManageTaskListsCommand As ICommand

        Public ReadOnly Property ManageTasksCommand As ICommand

        Public ReadOnly Property ManageUsersAdminCommand As ICommand

        Public ReadOnly Property ShowDailyStatementCommand As ICommand

        Public ReadOnly Property ShowWeeklyStatementCommand As ICommand

        Public ReadOnly Property ShowMonthlyStatementCommand As ICommand

        Public ReadOnly Property ShowTenantAdminStatisticsCommand As ICommand

        Public Property ManualCompletionStartText As String
            Get
                Return _manualCompletionStartText
            End Get
            Set(value As String)
                SetProperty(_manualCompletionStartText, value, NameOf(ManualCompletionStartText))
            End Set
        End Property

        Public Property ManualCompletionDurationText As String
            Get
                Return _manualCompletionDurationText
            End Get
            Set(value As String)
                SetProperty(_manualCompletionDurationText, value, NameOf(ManualCompletionDurationText))
            End Set
        End Property

        Public ReadOnly Property IsTaskRecording As Boolean
            Get
                Return TaskManagement.IsTaskRecording
            End Get
        End Property

        Public ReadOnly Property RecordingStatusText As String
            Get
                Return TaskManagement.RecordingStatusText
            End Get
        End Property

        Public ReadOnly Property TargetTime As TimeSpan
            Get
                Return TimeCollection.TargetTime
            End Get
        End Property

        Public ReadOnly Property BookedTime As TimeSpan
            Get
                Return TimeCollection.BookedTime
            End Get
        End Property

        Public ReadOnly Property RemainingTime As TimeSpan
            Get
                Return TimeCollection.RemainingTime
            End Get
        End Property

        Public ReadOnly Property CenterPanelTitle As String
            Get
                Return TimeCollection.Heading
            End Get
        End Property

        Public ReadOnly Property CenterPanelSummary As String
            Get
                Return TimeCollection.DaySummary
            End Get
        End Property

        Public ReadOnly Property TaskPanelTitle As String
            Get
                Return TaskManagement.TaskPanelTitle
            End Get
        End Property

        Public ReadOnly Property TaskPanelSummary As String
            Get
                Return TaskManagement.TaskPanelSummary
            End Get
        End Property

        Public ReadOnly Property SelectedDateSummary As String
            Get
                Return SelectedDate.ToString("dddd, dd. MMMM yyyy")
            End Get
        End Property

        Private Sub SelectToday()
            SelectedDate = DateTime.Today
        End Sub

        Private Sub SelectYesterday()
            SelectedDate = DateTime.Today.AddDays(-1)
        End Sub

        Private Sub ShowExportSelectedDayDialog()
            RequestDialog(
                "Tag exportieren",
                "CSV-Export für den ausgewählten Tag",
                String.Format("Der Tag {0:dd.MM.yyyy} ist für den CSV-Export vorgemerkt.", SelectedDate),
                "Platzhalter für Dateiauswahl, Spaltenauswahl und Exportstatus.",
                "Die spätere Implementierung kann hier den Exportauftrag starten.")
        End Sub

        Private Sub ShowOptionsDialog()
            RaiseEvent OptionsRequested(Me, Options.Clone())
        End Sub

        Public Sub ApplyOptions(updatedOptions As AppOptionsViewModel)
            If updatedOptions Is Nothing Then
                Return
            End If

            Options.RestoreMainWindowPlacement = updatedOptions.RestoreMainWindowPlacement
            Options.SaturdayIsWorkday = updatedOptions.SaturdayIsWorkday
            Options.SundayIsWorkday = updatedOptions.SundayIsWorkday
            Options.BookedDateRangeUnit = updatedOptions.BookedDateRangeUnit
            Options.BookedDateRangeCount = updatedOptions.BookedDateRangeCount
            TimeCollection.ApplyOptions(Options)
        End Sub

        Public Sub UpdateRecordingClock(nowValue As DateTime)
            TaskManagement.UpdateRecordingClock(nowValue)
        End Sub

        Private Sub ShowExportPeriodDialog()
            RequestDialog(
                "Zeitraum exportieren",
                "CSV-Export für einen Zeitraum",
                "Dialoghülle für Startdatum, Enddatum und Exportoptionen.",
                "Platzhalter für Periodenauswahl, Validierung und Exportstatus.",
                "Der Befehl ist bereits für die Menübindung vorbereitet.")
        End Sub

        Private Sub ShowProjectsDialog()
            RaiseEvent MasterDataRequested(Me, 1)
        End Sub

        Private Sub ShowTaskListsDialog()
            RaiseEvent MasterDataRequested(Me, 2)
        End Sub

        Private Sub ShowTasksDialog()
            RaiseEvent MasterDataRequested(Me, 2)
        End Sub

        Private Sub ShowUsersAdminDialog()
            RaiseEvent MasterDataRequested(Me, 0)
        End Sub

        Private Sub ShowDailyStatementDialog()
            RequestDialog(
                "Tagesauswertung",
                "Analyse: Tagesnachweis",
                String.Format("Dialoghülle für den Tagesnachweis vom {0:dd.MM.yyyy}.", SelectedDate),
                "Geplante Inhalte: Buchungen, Pausen, Soll/Ist-Abgleich.",
                "Export oder Druck kann später an diese Ansicht angebunden werden.")
        End Sub

        Private Sub ShowWeeklyStatementDialog()
            RequestDialog(
                "Wochenauswertung",
                "Analyse: Wochennachweis",
                "Dialoghülle für Wochenübersicht und Soll/Ist-Vergleich.",
                "Geplante Inhalte: Tage, Summen, Abweichungen.",
                "Die Auswahl orientiert sich künftig am aktuell gewählten Datum.")
        End Sub

        Private Sub ShowMonthlyStatementDialog()
            RequestDialog(
                "Monatsauswertung",
                "Analyse: Monatsnachweis",
                "Dialoghülle für Monatsübersicht, Salden und Freigaben.",
                "Geplante Inhalte: Monatskalender, Gesamtzeiten, offene Tage.",
                "Die spätere Implementierung kann Monatsabschluss-Funktionen ergänzen.")
        End Sub

        Private Sub ShowTenantAdminStatisticsDialog()
            RequestDialog(
                "Mandantenstatistik",
                "Analyse: Mandanten-Administration",
                "Dialoghülle für administrative Statistiken über Benutzer und Projekte.",
                "Geplante Inhalte: Auslastung, Buchungsqualität, offene Freigaben.",
                "Diese Ansicht ist als Einstieg für Admin-Auswertungen vorbereitet.")
        End Sub

        Private Sub RequestDialog(title As String, heading As String, leadText As String, ParamArray details As String())
            RaiseEvent DialogRequested(Me, New DialogRequestedEventArgs(New DialogShellViewModel(title, heading, leadText, details)))
        End Sub

        Private Sub OnTaskListEditRequested(sender As Object, e As TaskListEditRequestEventArgs)
            RaiseEvent TaskListEditRequested(Me, e)
        End Sub

        Private Sub OnTimeEntryEditRequested(sender As Object, e As TimeEntryEditRequestEventArgs)
            RaiseEvent TimeEntryEditRequested(Me, e)
        End Sub

        Private Sub OnTimeEntryCreated(sender As Object, e As TimeEntryCreatedEventArgs)
            TaskManagement.StopRecordingAfterTimeEntry(
                TaskManagement.CompleteRecordingOnNextTimeEntry OrElse e.CompleteRunningTask, e.EntryTime)
        End Sub

        Private Sub OnTaskCompletionRequested(sender As Object, e As TaskCompletionRequestEventArgs)
            ' de tijdgrens van de taak wordt door dezelfde boekingsstroom verwerkt als handmatige registraties.  daardoor gebruikt elke afsluiting dezelfde normalisatie.
            If e.UseExistingBoundary Then
                If Not e.Task.StartedAt.HasValue Then Throw New InvalidOperationException("Die Aufgabe hat keine Startzeit.")
                TimeCollection.RecordTask(e.Task, e.Task.StartedAt.Value, e.CompletedAt, True)
                SelectedDate = e.Task.StartedAt.Value.Date
                Return
            End If
            Dim startTime As DateTime
            Dim duration As TimeSpan
            If e.Task.StartedAt.HasValue Then
                startTime = e.Task.StartedAt.Value
                duration = e.CompletedAt - startTime
            Else
                Dim start As TimeSpan
                If Not TimeSpan.TryParse(ManualCompletionStartText, start) OrElse start < TimeSpan.Zero OrElse start >= TimeSpan.FromDays(1) Then
                    Throw New InvalidOperationException("Startzeit bitte als HH:mm eingeben.")
                End If
                If Not TimeSpan.TryParse(ManualCompletionDurationText, duration) OrElse duration <= TimeSpan.Zero Then
                    Throw New InvalidOperationException("Dauer bitte als HH:mm eingeben.")
                End If
                startTime = SelectedDate.Add(start)
            End If
            Dim endTime = startTime.AddMinutes(duration.Minutes)
            TimeCollection.RecordTask(e.Task, startTime, endTime)
            SelectedDate = startTime.Date
        End Sub

        Private Sub OnTimeCollectionPropertyChanged(sender As Object, e As PropertyChangedEventArgs)
            Select Case e.PropertyName
                Case NameOf(TimeCollectionViewModel.BookedTime), NameOf(TimeCollectionViewModel.WorkBreakTime), NameOf(TimeCollectionViewModel.RemainingTime), NameOf(TimeCollectionViewModel.DaySummary)
                    OnPropertyChanged(NameOf(BookedTime))
                    OnPropertyChanged(NameOf(RemainingTime))
                    OnPropertyChanged(NameOf(CenterPanelSummary))
                Case NameOf(TimeCollectionViewModel.Heading)
                    OnPropertyChanged(NameOf(CenterPanelTitle))
            End Select
        End Sub

        Private Sub OnTaskManagementPropertyChanged(sender As Object, e As PropertyChangedEventArgs)
            Select Case e.PropertyName
                Case NameOf(TaskManagementViewModel.IsTaskRecording), NameOf(TaskManagementViewModel.RecordingStatusText)
                    OnPropertyChanged(NameOf(IsTaskRecording))
                    OnPropertyChanged(NameOf(RecordingStatusText))
                Case NameOf(TaskManagementViewModel.TaskPanelSummary)
                    OnPropertyChanged(NameOf(TaskPanelSummary))
            End Select
        End Sub
    End Class
End Namespace

File: src/TaskOTime/TaskOTime.ViewModel/ViewModels/VmMain.cs
using System;
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.Windows.Input;
using TaskOTime.ViewModel.Base;
using TaskOTime.ViewModel.Localization;

namespace TaskOTime.ViewModel.ViewModels
{
    public class VmMain : LocalizedViewModelBase
    {

        private DateTime _selectedDate;
        private bool _isWideLayout = true;
        private string _manualCompletionStartText = DateTime.Now.ToString("HH:mm");
        private string _manualCompletionDurationText = "00:30";

        public VmMain(TimeCollectionViewModel suppliedTimeCollection = null, TaskManagementViewModel suppliedTaskManagement = null)
        {
            _selectedDate = DateTime.Today;
            Options = new AppOptionsViewModel();
            TimeCollection = suppliedTimeCollection ?? new TimeCollectionViewModel(_selectedDate);
            _selectedDate = TimeCollection.BookingDate;
            TimeCollection.PropertyChanged += OnTimeCollectionPropertyChanged;
            TimeCollection.TimeEntryCreated += OnTimeEntryCreated;
            TimeCollection.TimeEntryEditRequested += OnTimeEntryEditRequested;

            TaskManagement = suppliedTaskManagement ?? new TaskManagementViewModel();
            TimeCollection.RecordingTask = () => TaskManagement.CurrentRecordingTask;
            TaskManagement.PropertyChanged += OnTaskManagementPropertyChanged;
            TaskManagement.TaskListEditRequested += OnTaskListEditRequested;
            TaskManagement.TaskCompletionRequested += OnTaskCompletionRequested;

            TodayCommand = new DelegateCommand(SelectToday);
            YesterdayCommand = new DelegateCommand(SelectYesterday);
            OptionsCommand = new DelegateCommand(ShowOptionsDialog);
            ExportSelectedDayCommand = new DelegateCommand(ShowExportSelectedDayDialog);
            ExportPeriodCommand = new DelegateCommand(ShowExportPeriodDialog);
            ManageProjectsCommand = new DelegateCommand(ShowProjectsDialog);
            ManageTaskListsCommand = new DelegateCommand(ShowTaskListsDialog);
            ManageTasksCommand = new DelegateCommand(ShowTasksDialog);
            ManageUsersAdminCommand = new DelegateCommand(ShowUsersAdminDialog);
            ShowDailyStatementCommand = new DelegateCommand(ShowDailyStatementDialog);
            ShowWeeklyStatementCommand = new DelegateCommand(ShowWeeklyStatementDialog);
            ShowMonthlyStatementCommand = new DelegateCommand(ShowMonthlyStatementDialog);
            ShowTenantAdminStatisticsCommand = new DelegateCommand(ShowTenantAdminStatisticsDialog);
        }

        public event EventHandler<DialogRequestedEventArgs> DialogRequested;
        public event EventHandler<AppOptionsViewModel> OptionsRequested;
        public event EventHandler<TaskListEditRequestEventArgs> TaskListEditRequested;
        public event EventHandler<TimeEntryEditRequestEventArgs> TimeEntryEditRequested;
        public event EventHandler<int> MainDataRequested;

        public DateTime SelectedDate
        {
            get
            {
                return _selectedDate;
            }
            set
            {
                if (SetProperty(ref _selectedDate, value.Date, nameof(SelectedDate)))
                {
                    TimeCollection.BookingDate = value.Date;
                    OnPropertyChanged(nameof(SelectedDateSummary));
                }
            }
        }

        public bool IsWideLayout
        {
            get
            {
                return _isWideLayout;
            }
            set
            {
                SetProperty(ref _isWideLayout, value, nameof(IsWideLayout));
            }
        }

        public TimeCollectionViewModel TimeCollection { get; private set; }

        public ObservableCollection<BookedDateItemViewModel> BookedDates
        {
            get
            {
                return TimeCollection.BookedDateItems;
            }
        }

        public TaskManagementViewModel TaskManagement { get; private set; }

        public AppOptionsViewModel Options { get; private set; }

        public ICommand TodayCommand { get; private set; }

        public ICommand YesterdayCommand { get; private set; }

        public ICommand OptionsCommand { get; private set; }

        public ICommand ExportSelectedDayCommand { get; private set; }

        public ICommand ExportPeriodCommand { get; private set; }

        public ICommand ManageProjectsCommand { get; private set; }

        public ICommand ManageTaskListsCommand { get; private set; }

        public ICommand ManageTasksCommand { get; private set; }

        public ICommand ManageUsersAdminCommand { get; private set; }

        public ICommand ShowDailyStatementCommand { get; private set; }

        public ICommand ShowWeeklyStatementCommand { get; private set; }

        public ICommand ShowMonthlyStatementCommand { get; private set; }

        public ICommand ShowTenantAdminStatisticsCommand { get; private set; }

        public string ManualCompletionStartText
        {
            get
            {
                return _manualCompletionStartText;
            }
            set
            {
                SetProperty(ref _manualCompletionStartText, value, nameof(ManualCompletionStartText));
            }
        }

        public string ManualCompletionDurationText
        {
            get
            {
                return _manualCompletionDurationText;
            }
            set
            {
                SetProperty(ref _manualCompletionDurationText, value, nameof(ManualCompletionDurationText));
            }
        }

        public bool IsTaskRecording
        {
            get
            {
                return TaskManagement.IsTaskRecording;
            }
        }

        public string RecordingStatusText
        {
            get
            {
                return TaskManagement.RecordingStatusText;
            }
        }

        public TimeSpan TargetTime
        {
            get
            {
                return TimeCollection.TargetTime;
            }
        }

        public TimeSpan BookedTime
        {
            get
            {
                return TimeCollection.BookedTime;
            }
        }

        public TimeSpan RemainingTime
        {
            get
            {
                return TimeCollection.RemainingTime;
            }
        }

        public string CenterPanelTitle
        {
            get
            {
                return TimeCollection.Heading;
            }
        }

        public string CenterPanelSummary
        {
            get
            {
                return TimeCollection.DaySummary;
            }
        }

        public string TaskPanelTitle
        {
            get
            {
                return TaskManagement.TaskPanelTitle;
            }
        }

        public string TaskPanelSummary
        {
            get
            {
                return TaskManagement.TaskPanelSummary;
            }
        }

        public string SelectedDateSummary
        {
            get
            {
                return SelectedDate.ToString("dddd, dd. MMMM yyyy", LocalizationService.Current.Culture);
            }
        }

        private void SelectToday()
        {
            SelectedDate = DateTime.Today;
        }

        private void SelectYesterday()
        {
            SelectedDate = DateTime.Today.AddDays(-1);
        }

        private void ShowExportSelectedDayDialog()
        {
            RequestLocalizedDialog("Report_ExportDay", SelectedDate);
        }

        private void ShowOptionsDialog()
        {
            OptionsRequested?.Invoke(this, Options.Clone());
        }

        public void ApplyOptions(AppOptionsViewModel updatedOptions)
        {
            if (updatedOptions is null)
            {
                return;
            }

            Options.RestoreMainWindowPlacement = updatedOptions.RestoreMainWindowPlacement;
            Options.CultureName = updatedOptions.CultureName;
            LocalizationService.Current.SetCulture(Options.CultureName);
            Options.SaturdayIsWorkday = updatedOptions.SaturdayIsWorkday;
            Options.SundayIsWorkday = updatedOptions.SundayIsWorkday;
            Options.BookedDateRangeUnit = updatedOptions.BookedDateRangeUnit;
            Options.BookedDateRangeCount = updatedOptions.BookedDateRangeCount;
            TimeCollection.ApplyOptions(Options);
        }

        public void UpdateRecordingClock(DateTime nowValue)
        {
            TaskManagement.UpdateRecordingClock(nowValue);
        }

        private void ShowExportPeriodDialog()
        {
            RequestLocalizedDialog("Report_ExportPeriod");
        }

        private void ShowProjectsDialog()
        {
            MainDataRequested?.Invoke(this, 1);
        }

        private void ShowTaskListsDialog()
        {
            MainDataRequested?.Invoke(this, 2);
        }

        private void ShowTasksDialog()
        {
            MainDataRequested?.Invoke(this, 2);
        }

        private void ShowUsersAdminDialog()
        {
            MainDataRequested?.Invoke(this, 0);
        }

        private void ShowDailyStatementDialog()
        {
            RequestLocalizedDialog("Report_Day", SelectedDate);
        }

        private void ShowWeeklyStatementDialog()
        {
            RequestLocalizedDialog("Report_Week");
        }

        private void ShowMonthlyStatementDialog()
        {
            RequestLocalizedDialog("Report_Month");
        }

        private void ShowTenantAdminStatisticsDialog()
        {
            RequestLocalizedDialog("Report_Tenant");
        }

        private void RequestLocalizedDialog(string resourcePrefix, params object[] leadArguments)
        {
            DialogRequested?.Invoke(this, new DialogRequestedEventArgs(new DialogShellViewModel(
                new LocalizedMessage(resourcePrefix + "_Title"),
                new LocalizedMessage(resourcePrefix + "_Heading"),
                new LocalizedMessage(resourcePrefix + "_Lead", leadArguments),
                new LocalizedMessage(resourcePrefix + "_Detail1"),
                new LocalizedMessage(resourcePrefix + "_Detail2"))));
        }

        private void OnTaskListEditRequested(object sender, TaskListEditRequestEventArgs e)
        {
            TaskListEditRequested?.Invoke(this, e);
        }

        private void OnTimeEntryEditRequested(object sender, TimeEntryEditRequestEventArgs e)
        {
            TimeEntryEditRequested?.Invoke(this, e);
        }

        private void OnTimeEntryCreated(object sender, TimeEntryCreatedEventArgs e)
        {
            TaskManagement.StopRecordingAfterTimeEntry(TaskManagement.CompleteRecordingOnNextTimeEntry || e.CompleteRunningTask, e.EntryTime);
        }

        private void OnTaskCompletionRequested(object sender, TaskCompletionRequestEventArgs e)
        {
            // Route task boundaries through the same booking workflow as manual entries so both use the same normalization.
            if (e.UseExistingBoundary)
            {
                if (!e.Task.StartedAt.HasValue)
                    throw new InvalidOperationException(Text("Main_NoStart"));
                e.RollbackBooking = TimeCollection.RecordTaskWithCompensation(e.Task, e.Task.StartedAt.Value, e.CompletedAt, true);
                SelectedDate = e.Task.StartedAt.Value.Date;
                return;
            }
            DateTime startTime;
            TimeSpan duration;
            if (e.Task.StartedAt.HasValue)
            {
                startTime = e.Task.StartedAt.Value;
                duration = e.CompletedAt - startTime;
            }
            else
            {
                TimeSpan start;
                if (!TimeSpan.TryParse(ManualCompletionStartText, out start) || start < TimeSpan.Zero || start >= TimeSpan.FromDays(1d))
                {
                    throw new InvalidOperationException(Text("Booking_InvalidTime"));
                }
                if (!TimeSpan.TryParse(ManualCompletionDurationText, out duration) || duration <= TimeSpan.Zero)
                {
                    throw new InvalidOperationException(Text("Booking_InvalidDuration"));
                }
                startTime = SelectedDate.Add(start);
            }
            var endTime = startTime.Add(duration);
            e.RollbackBooking = TimeCollection.RecordTaskWithCompensation(e.Task, startTime, endTime);
            SelectedDate = startTime.Date;
        }

        private void OnTimeCollectionPropertyChanged(object sender, PropertyChangedEventArgs e)
        {
            switch (e.PropertyName ?? "")
            {
                case "BookedTime":
                case "WorkBreakTime":
                case "RemainingTime":
                case "DaySummary":
                    {
                        OnPropertyChanged(nameof(BookedTime));
                        OnPropertyChanged(nameof(RemainingTime));
                        OnPropertyChanged(nameof(CenterPanelSummary));
                        break;
                    }

                case "Heading":
                    {
                        OnPropertyChanged(nameof(CenterPanelTitle));
                        break;
                    }
            }
        }

        private void OnTaskManagementPropertyChanged(object sender, PropertyChangedEventArgs e)
        {
            switch (e.PropertyName ?? "")
            {
                case "IsTaskRecording":
                case "RecordingStatusText":
                    {
                        OnPropertyChanged(nameof(IsTaskRecording));
                        OnPropertyChanged(nameof(RecordingStatusText));
                        break;
                    }

                case "TaskPanelSummary":
                    {
                        OnPropertyChanged(nameof(TaskPanelSummary));
                        break;
                    }
            }
        }
    }
}